flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,412 @@
+            if (map.size() < topN) {
+                map.add(Tuple2.of(newValue.f0, newValue.f1));
+                head = map.peek();
+            } else {
+                if (map.comparator().compare(head, newValue) < 0) {
+                    Tuple2<Double, T> peek = map.poll();
+                    assert peek != null;
+                    peek.f0 = newValue.f0;
+                    peek.f1 = newValue.f1;
+                    map.add(peek);
+                    head = map.peek();
+                }
+            }
+            return head;
+        }
+
+        /**
+         * Computes distance between sample and dictionary vectors.
+         *
+         * @param input sample with l2 norm.
+         * @param index dictionary vectors index.
+         * @return distances.
+         */
+        private ArrayList<Tuple2<Double, Object>> computeDistance(
+                Tuple2<DenseVector, Double> input, Integer index) {
+            Tuple3<DenseMatrix, DenseVector, String[]> data = modelData.getDictData().get(index);
+
+            DenseMatrix res = calc(input, data);
+            ArrayList<Tuple2<Double, Object>> list = new ArrayList<>(0);
+            String[] curLabels = data.f2;
+            for (int i = 0; i < Objects.requireNonNull(curLabels).length; i++) {
+                Tuple2<Double, Object> tuple = Tuple2.of(res.values[i], curLabels[i]);
+                list.add(tuple);
+            }
+            return list;
+        }
+
+        /** The blas used to accelerating speed. */
+        private static final dev.ludovic.netlib.blas.F2jBLAS NATIVE_BLAS =
+                (F2jBLAS) F2jBLAS.getInstance();
+
+        /**
+         * Compute distance between sample and dictionary vectors.
+         *
+         * @param left Sample and norm.
+         * @param right Dictionary vectors with row format.
+         * @return a new DenseMatrix which store the result distance.
+         */
+        public DenseMatrix calc(
+                Tuple2<DenseVector, Double> left,
+                Tuple3<DenseMatrix, DenseVector, String[]> right) {
+            DenseMatrix vectors = right.f0;
+            DenseMatrix res = new DenseMatrix(Objects.requireNonNull(vectors).numCols, 1);
+            DenseVector norm = right.f1;
+            double[] normL2Square = Objects.requireNonNull(norm).values;
+
+            final int m = vectors.numRows;
+            final int n = vectors.numCols;
+            NATIVE_BLAS.dgemv(
+                    "T", m, n, -2.0, vectors.values, m, left.f0.toArray(), 1, 0.0, res.values, 1);
+
+            for (int i = 0; i < res.values.length; i++) {
+                res.values[i] = Math.sqrt(Math.abs(res.values[i] + left.f1 + normL2Square[i]));
+            }
+            return res;
+        }
+
+        /**
+         * Computes norm2 of vector.
+         *
+         * @return Sample with norm2.
+         */
+        public static Tuple2<DenseVector, Double> computeNorm(DenseVector vector) {
```

Comment:
Could this method be private? Same for other method such as `calc(...)` and `map(...)` defined in this file.

Reply:
OK

Reply:
done

Reply:
done